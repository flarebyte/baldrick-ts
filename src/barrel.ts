import { readFileSync, writeFileSync } from 'fs';
import isEqual from 'lodash.isequal';

export { readFileSync, writeFileSync, isEqual };
