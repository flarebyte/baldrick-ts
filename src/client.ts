import { commanding } from './index.js';

/**
 * This function may be merged in the future when the linter does a better job at recognizing .mts files
 */
export async function runClient() {
  try {
    await commanding.parseAsyncArgv();
  } catch (err) {
    console.log('baldrick-ts generator will exit with error code 1');
    console.error(err);
    process.exit(1);
  }
}
