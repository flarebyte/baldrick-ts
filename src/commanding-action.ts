import { updateAll } from './io-sfx.js';
import { GenerateAction, GenerateActionOpts, RunnerContext } from './model.js';

export const cmdGenerateAction: GenerateAction = async (
  ctx: RunnerContext,
  options: GenerateActionOpts
) => {
  await updateAll(ctx, options);
  ctx.termFormatter({
    title: 'Normalized the project structure and documents',
    detail: ['* Search for any fixme in your text editor'].join('\n'),
    kind: 'info',
    format: 'default',
  });
};
