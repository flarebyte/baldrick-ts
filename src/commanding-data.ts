import { CmdOption, CmdOptionsGenerator } from './model.js';

const githubAccount: CmdOption = {
  shortFlag: 'ga',
  longFlag: 'github-account',
  description: 'Github account',
  choices: [],
  mandatory: true,
  variadic: false,
};

const name: CmdOption = {
  shortFlag: 'n',
  longFlag: 'name',
  description: 'Name of the library',
  choices: [],
  mandatory: false,
  variadic: false,
};

const bin: CmdOption = {
  shortFlag: 'b',
  longFlag: 'bin',
  description: 'The binary command used for calling the CLI',
  choices: [],
  mandatory: false,
  variadic: false,
};

const feature: CmdOption = {
  shortFlag: 'f',
  longFlag: 'feature',
  description: 'List of features',
  choices: ['lib', 'cli', 'gen'],
  mandatory: true,
  variadic: true,
};

const license: CmdOption = {
  shortFlag: 'l',
  longFlag: 'license',
  description: 'Open source license if any',
  defaultValue: 'MIT',
  choices: ['MIT', 'UNLICENSED'],
  mandatory: false,
  variadic: false,
};

const copyrightHolder: CmdOption = {
  shortFlag: 'ch',
  longFlag: 'copyright-holder',
  description: 'The owner of the copyright regarding the source code',
  choices: [],
  mandatory: false,
  variadic: false,
};

const copyrightStartYear: CmdOption = {
  shortFlag: 'cy',
  longFlag: 'copyright-start-year',
  description: 'The year the project has been started',
  defaultValue: '2021',
  choices: [],
  mandatory: false,
  variadic: false,
};

export const cmdOptionsGenerator: CmdOptionsGenerator = {
  feature,
  name,
  githubAccount,
  license,
  copyrightHolder,
  copyrightStartYear,
  bin,
};
