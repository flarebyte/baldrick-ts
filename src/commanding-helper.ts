import { Option } from 'commander';
import { CmdOption } from './model.js';

const capitalize = (value: string): string =>
  value.length > 0 ? (value[0] || '').toUpperCase() + value.slice(1) : '';

const decapitalize = (value: string): string =>
  value.length > 0 ? (value[0] || '').toLowerCase() + value.slice(1) : '';

export const toCamelCase = (longFlag: string): string =>
  decapitalize(longFlag.split('-').map(capitalize).join(''));

export const toCommanderOption = (option: CmdOption): Option => {
  const dot3 = option.variadic ? '...' : '';
  const flags = `-${option.shortFlag}, --${option.longFlag} [${toCamelCase(
    option.longFlag
  )}${dot3}]`;
  const opts = new Option(flags, option.description);
  opts.defaultValue = option.defaultValue;
  if (option.choices.length > 0) {
    opts.choices(option.choices);
  }
  opts.mandatory = option.mandatory;
  return opts;
};
