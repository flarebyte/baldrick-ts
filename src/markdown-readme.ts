import { markdownToString, parseMarkdown } from './markdown.js';
import { Badge, CoreProject, MdSection } from './model.js';

const capitalize = (value: string) =>
  (value[0]?.toUpperCase() || '') + value.slice(1);

const libBadges = (core: CoreProject): Badge[] => {
  const always: Badge[] = [
    {
      text: 'npm',
      imageUrl: `https://img.shields.io/npm/v/${core.name}`,
      position: 'top',
    },
    {
      text: 'Build status',
      imageUrl: `https://github.com/${core.githubAccount}/${core.name}/actions/workflows/main.yml/badge.svg`,
      position: 'top',
    },
    {
      text: 'npm bundle size',
      imageUrl: `https://img.shields.io/bundlephobia/min/${core.name}`,
      position: 'top',
    },
    {
      text: 'npm type definitions',
      imageUrl: `https://img.shields.io/npm/types/${core.name}`,
      position: 'bottom',
    },
    {
      text: 'node-current',
      imageUrl: `https://img.shields.io/node/v/${core.name}`,
      position: 'bottom',
    },
    {
      text: 'NPM',
      imageUrl: `https://img.shields.io/npm/l/${core.name}`,
      position: 'bottom',
    },
  ];
  const codacy: Badge[] = core.codacyId
    ? [
        {
          text: 'Codacy Badge',
          imageUrl: `https://app.codacy.com/project/badge/Grade/${core.codacyId}`,
          position: 'top',
          refUrl: `https://www.codacy.com/gh/${core.githubAccount}/${core.name}/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=${core.githubAccount}/${core.name}&amp;utm_campaign=Badge_Grade`,
        },
      ]
    : [];
  return [...always, ...codacy];
};

const keepSections = (section: MdSection): boolean =>
  ['Usage', 'Advanced use', 'Acknowledgements'].includes(section.title);

const docAndLinks = (core: CoreProject): MdSection => ({
  title: 'Documentation and links',
  body: [
    '* [Code Maintenance](MAINTENANCE.md)',
    '* [Code Of Conduct](CODE_OF_CONDUCT.md)',
    `* [Api for ${core.name}](API.md)`,
    '* [Contributing](CONTRIBUTING.md)',
    `* [Contributors](https://github.com/${core.githubAccount}/${core.name}/graphs/contributors)`,
    `* [Dependencies](https://github.com/${core.githubAccount}/${core.name}/network/dependencies)`,
  ].join('\n'),
});

const installSection = (core: CoreProject): MdSection => {
  const isCli = core.feature.includes('cli');
  const bodyLib = ['```', `yarn add ${core.name}`, '```'];
  const bodyCli = [
    '```bash',
    `yarn global add ${core.name}`,
    `${core.bin} --help`,
    '```',
    'Or alternatively run it:',
    '```bash',
    `npx ${core.name} --help`,
    '```',
    'If you want to tun the latest version from github. Mostly useful for dev:',
    '```bash',
    `git clone git@github.com:${core.githubAccount}/${core.name}.git`,
    'yarn global add `pwd`',
    '```',
  ];
  const body = isCli ? bodyCli : bodyLib;

  return {
    title: 'Installation',
    body: [
      'This package is [ESM only](https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77).',
      ...body,
    ].join('\n'),
  };
};

export const toReadmeMd = (core: CoreProject, existingMd: string): string => {
  const existing = parseMarkdown(existingMd);
  const title = capitalize(core.name);
  const badges = libBadges(core);
  const sections = [
    ...existing.sections.filter(keepSections),
    docAndLinks(core),
    installSection(core),
  ];
  const updated = {
    ...existing,
    title,
    badges,
    sections,
  };
  const rawReadme = markdownToString(updated);
  return rawReadme.replace(/[\n\r]{3,}/gm, '\n\n');
};
