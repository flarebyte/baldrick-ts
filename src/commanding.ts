import { Command } from 'commander';
import { cmdOptionsGenerator } from './commanding-data.js';
import { toCommanderOption } from './commanding-helper.js';
import { toFeatures } from './feature-helper.js';
import {
  GenerateAction,
  GenerateActionOpts,
  GenerateRawOpts,
  RunnerContext,
} from './model.js';
import { basicFormatter, errorFormatter } from './term-formatter.js';
import { version } from './version.js';

export class Commanding {
  _program: Command = new Command();
  constructor() {
    this._program.version(version);
  }

  declareGenerateAction(genAction: GenerateAction) {
    this._program
      .command('generate')
      .description('Generate and standardize the source code files')
      .addOption(toCommanderOption(cmdOptionsGenerator.feature))
      .addOption(toCommanderOption(cmdOptionsGenerator.name))
      .addOption(toCommanderOption(cmdOptionsGenerator.githubAccount))
      .addOption(toCommanderOption(cmdOptionsGenerator.copyrightHolder))
      .addOption(toCommanderOption(cmdOptionsGenerator.copyrightStartYear))
      .addOption(toCommanderOption(cmdOptionsGenerator.license))
      .addOption(toCommanderOption(cmdOptionsGenerator.bin))
      .addOption(toCommanderOption(cmdOptionsGenerator.codacyId))
      .action(async (options: GenerateRawOpts) => {
        const {
          feature,
          name,
          githubAccount,
          copyrightHolder,
          license,
          copyrightStartYear,
          bin,
          codacyId,
        } = options;
        const generateOpts: GenerateActionOpts = {
          feature: toFeatures(feature),
          name,
          githubAccount,
          copyrightHolder,
          license,
          copyrightStartYear: Number.parseInt(copyrightStartYear),
          bin,
          codacyId,
        };
        const ctx: RunnerContext = {
          currentPath: process.cwd(),
          currentYear: new Date().getFullYear(),
          termFormatter: basicFormatter,
          errTermFormatter: errorFormatter,
        };
        await genAction(ctx, generateOpts);
      });
  }

  async parseAsync(argv: string[]) {
    return await this._program.parseAsync(argv, { from: 'node' });
  }

  async parseAsyncArgv() {
    return await this.parseAsync(process.argv);
  }
}
