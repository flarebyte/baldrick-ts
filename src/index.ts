import { cmdGenerateAction } from './commanding-action.js';
import { Commanding } from './commanding.js';

const commanding = new Commanding();

commanding.declareGenerateAction(cmdGenerateAction);

export { commanding };
