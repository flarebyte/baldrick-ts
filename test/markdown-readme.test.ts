import { libCoreProject } from './fixture-core-project';
import { toReadmeMd } from '../src/markdown-readme';
describe('Readme documentation', () => {
  it('Updates README.md with standardized chapters', () => {
    const existingReadme = `
        # About this doc
        `;
    const actual = toReadmeMd(libCoreProject, existingReadme);
    expect(actual).toBeDefined();
  });
});
