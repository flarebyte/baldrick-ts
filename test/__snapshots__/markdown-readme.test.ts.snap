// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Readme documentation Updates README.md with standardized chapters 1`] = `
"# Project123

![npm](https://img.shields.io/npm/v/project123)

![Build status](https://github.com/mycompany/project123/actions/workflows/main.yml/badge.svg)

![npm bundle size](https://img.shields.io/bundlephobia/min/project123)

![Dependencies](https://status.david-dm.org/gh/mycompany/project123.svg)

![npm type definitions](https://img.shields.io/npm/types/project123)

![node-current](https://img.shields.io/node/v/project123)

![NPM](https://img.shields.io/npm/l/project123)

> Scratch Book is an annotation library written in Typescript that makes it easy to write documentation in a decentralized fashion.

- **Typescript** can be used to write the documentation. Integrates perfectly with [ts-node](https://typestrong.org/ts-node/).
- **No dependency** and lightweight.
- **Advanced matchers** to query the notes.
- **Simple model** for the notes that requires very little learning curve while still offering a lot of flexibility.
- **extensive tests coverage**.

## Usage

\`\`\`
const myScratchBook: ScratchBook = {
notes: [
createScratchNote('about/scratchbook/description', 'Annotation library'),
createScratchNote('about/scratchbook/github', 'scratchbook on github', 'https://github.com/flarebyte/scratchbook'),
createScratchNote('about/other/...', 'Some other projects')
]
}
filterScratchBook(withPrefix('about/scratchbook/'))(myScratchBook)
// will return the description and the github link

\`\`\`

### Documentation

A detailed documentation for each functions is [available](https://flarebyte.github.io/scratchbook/)

### Installation

\`\`\`
yarn add scratchbook
\`\`\`

## Documentation and links

* [Code Maintenance](MAINTENANCE.md)
* [Code Of Conduct](CODE_OF_CONDUCT.md)
* [Contributing](CONTRIBUTING.md)
* [Contributors](https://github.com/mycompany/project123/graphs/contributors)
* [Dependencies](https://github.com/mycompany/project123/network/dependencies)

## Installation

This package is [ESM only][esm].
\`\`\`
yarn add project123
\`\`\`"
`;
